godot-core/src/builtin/string/node_path.rs

```diff
@@ -49,6 +80,50 @@ impl NodePath {
             .expect("Godot hashes are uint32_t")
     }
 
+    /// Returns the property name indicated by `idx`, starting from 0. If `idx` is out of bounds, [`None`] is returned.
+    /// See also [`NodePath::get_subname_count`].
+    pub fn get_subname(&self, idx: u32) -> Option<StringName> {
+        let inner = self.as_inner();
+        let idx = idx as i64;
+        // This check checks both data being empty (get_subname_count returns 0) and index out of bounds.
+        if idx >= inner.get_subname_count() {
+            None
+        } else {
+            Some(inner.get_subname(idx))
+        }
+    }
+
+    /// Returns the slice of the [`NodePath`] as a new [`NodePath`]
+    pub fn slice(&self, range: impl RangeBounds<i64>) -> NodePath {
+        self.as_inner().slice(
+            match range.start_bound() {
+                Excluded(&start) => {
+                    if start == -1 {
+                        // Default end from godot, since the start is excluded.
+                        i32::MAX as i64
+                    } else {
+                        start + 1
+                    }
+                }
+                Included(&start) => start,
+                Unbounded => 0,
+            },
+            match range.end_bound() {
+                Excluded(&end) => end,
+                Included(&end) => {
+                    if end == -1 {
+                        // Default end from godot, since the end is excluded.
+                        i32::MAX as i64
+                    } else {
+                        end + 1
+                    }
+                }
+                // Default end from godot.
+                Unbounded => i32::MAX as i64,
+            },
+        )
+    }
```

Comment:
As mentioned on Discord, I'm not sure if we should provide ranges _and_ negative indices. These two are usually exclusive, and I'd say Rust users expect ranges to be monotonically increasing.

GDScript uses [this signature](https://docs.godotengine.org/en/stable/classes/class_nodepath.html#class-nodepath-method-slice):
```gdscript
NodePath slice(begin: int, end: int = 2147483647) const 
```

If you don't want to provide `Option<i64>` for the `end` parameter, you can always have two separate functions `slice` and `slice_from` or so.

Reply:
Alternatively, we don't provide negative indices, in which case you might be able to reuse `to_fromlen_pair` from my PR https://github.com/godot-rust/gdext/pull/980.

Reply:
Wouldn't it be better to specify in the documentation for the function that the functionality is there? My points in favor of adding it are the following:

- Direct portability from Godot, not losing any of the functionality
- Positive ranges are still usable, adding negative indices is only an extra, it doesn't take away functionality from positive ranges.
- Ranges allow this behavior, so it's not anti-Rust.
- In this case, changing a negative range to a positive range is not trivial, one would need to use `get_names_count()-i`, which is not as simple as your usual len function.

For all of the above I'd advocate to keep the ranges in negative (using isize instead), but if not possible, I'd use either the Option<> one or add a new function named slice_from for the one with only start. I think both the Option and the slice from are more cumbersome than the range approach though. After your decision I'll change the code.

Reply:
I don't think this API is used nearly enough to deserve such a lengthy discussion :sweat_smile: but since we're here already...

In my opinion, consistency and principle of least surprise are important API design guidelines. There aren't any other places in godot-rust where negative ranges are used as wrap-around. If we allow these semantics here, people may be surprised why it doesn't work in other places that take ranges, e.g. `substr` or in the future index operators taking ranges.

There is also the precedent of [`Array::subarray_{deep|shallow}`](https://godot-rust.github.io/docs/gdext/master/godot/builtin/struct.Array.html#method.subarray_shallow) and [`Packed*Array::subarray()` ](https://godot-rust.github.io/docs/gdext/master/godot/builtin/struct.PackedInt32Array.html#method.subarray), which are all called `slice` in Godot, and currently take `usize`.

We might reconsider for them to take `i64` to allow negative indices like here.

Btw, maybe this method should also be called `subpath` instead of `slice`, with a doc alias :thinking:

Reply:
> In my opinion, consistency and principle of least surprise are important API design guidelines.

Yes, absolutely, totally agree to this, no questions asked. I will follow precedent and if it's decided it should be changed, I'll change it myself if asked.

> There aren't any other places in godot-rust where negative ranges are used as wrap-around.

Then, if it's ever considered, it shouldn't be that hard to change (the commit is here for easy copy paste)

> There is also the precedent of Array::subarray_{deep|shallow} and Packed*Array::subarray() , which are all called slice in Godot, and currently take usize.

I would absolutely advocate for them taking isize instead. For the mentioned earlier, i64 is not intuitive for indices or slices, isize I feel makes more sense. It would provide feature parity with Godot too. For now, I'll do exactly as those methods and use usize and clamp the value to the length. No default value for end as it is not provided in these methods. To do it, I'd use the precedent from deep/shallow and make subpath and subpath_from, at most.

> Btw, maybe this method should also be called subpath instead of slice, with a doc alias 🤔

Agreed, I don't know the API for the doc alias, so if you could point me to that, I'll use it.

Reply:
While `isize` can make more sense in theory, it's not a type that Godot uses, and thus none of our API has it (back to consistency :slightly_smiling_face: ).

Looking at generated classes, some of them with "index" semantics use `i32`:
- https://godot-rust.github.io/docs/gdext/master/godot/classes/struct.Node.html#method.get_index
- https://godot-rust.github.io/docs/gdext/master/godot/classes/struct.ArrayMesh.html
- https://godot-rust.github.io/docs/gdext/master/godot/classes/struct.SurfaceTool.html#method.add_index
- https://godot-rust.github.io/docs/gdext/master/godot/classes/native_menu/struct.ExAddIconCheckItem.html#method.index
- https://godot-rust.github.io/docs/gdext/master/godot/classes/text_edit/struct.ExGetLineWidth.html#method.wrap_index

A few use `u32` :see_no_evil: 
- https://godot-rust.github.io/docs/gdext/master/godot/classes/struct.RenderingDevice.html#method.index_array_create

This is something we need to live with. The Godot API is not perfect, and `usize/isize` aren't types that GDExtension uses. It's a bit hard to assume that conversions in the general case are always safe; a method taking `i32` could in fact allow negative values, so we can't just assume `usize` fits and replace everywhere automatically. (Even though it would be nice... but it would take more effort)

But good that you brought this up, it shows that `i64` is also very uncommon for index-like things. Overall we should reduce the number of types needed to a minimum. Since both `isize` and `i64` are uncommon, let's stick to the ones with precedent:
- `usize` for pure indices/offsets that are guaranteed to be positive (e.g. `Array` has those)
- `i32` for indices that can either be negative (wrap-around), or whose semantics aren't so clear-cut

Maybe the `subarray*` methods should also be changed to take `i32`? (Not in this PR, since that's a breaking change) :thinking:

Reply:
Can't say I like the i32's choice, it feels as arbitrary as the i64, specially when using size for the unsigned and 32 for the signed. I stand on the side that using sizes for everything is the most intuitive, or at least, signed version should use double the bits (thus i64) since the signed part and the unsigned both refer to the same. That said, first is consistency and precedent is good to follow, so I understand the decision.

I agree that subarrays should use the signed version going forward for parity and usefulness.

Should I code the subpath already in i32 instead then? Just to confirm, I was going with usize to stick to subarray precedent

Reply:
I think `i32` is OK if we want to support the negative wrap-around. We can then later adjust `subarray`  to match this.

Agree that `i32` isn't the best, but introducing `isize` would mean either of:
- having yet another type to which the user needs to convert (there's no reason why exactly _this_ API should be different)
- if we stay consistent, change generated Godot APIs, which is an option but a bigger endeavor

Reply:
> I think i32 is OK

Then I'll implement it in i32 using clamps for code consistency.

> - having yet another type to which the user needs to convert (there's no reason why exactly this API should be different)

I agree that this is just bad and confusing, that's why I'd only choose this type if all "negative indexing" changed with it.

> - if we stay consistent, change generated Godot APIs, which is an option but a bigger endeavor

My issue with even thinking of following Godot APIs is how they're either misleading at best (i64 when using u32 for hashes, for example) or error prone and wrong at best. They say that the end defaults to i32::MAX yet I believe the sizes of the path, subpath, and whatnot are not i32 but simply int. That's why so long as we do clamping (which is a code pattern in both subarray methods, for example) we should be safe in using what from a user standpoint makes the most sense from Rust (in the case of sizes, indexes and whatnot, usize and isize). But I acknowledge it is quite the endeavor + breaking change and this is just my perspective that it has changed from the start of the conversation to now, so, I don't have much ground to hold